### ISSUE TITLE
How to retain reference information in Stringtie --merge

### ISSUE THREAD

**Issue by Michael-w-s** [gpertea/stringtie#79]
Hi,

I have many samples from multiple tissues that I am running against one reference and would like to retain reference information such as the gene name. Is there a way to do this or to force Stringtie to only accept genes found in the reference? I used -B -e and -G options but it still seems to be generating novel genes or isoforms.

**Comment by gpertea** [comment]
Not sure what you're asking and your last statement seems inaccurate, because when used with -e StringTie does not generate any novel genes or isoforms, so perhaps the actual "reference" you used for the -G option had those already?
Perhaps what you want is to just run the "shortcut" pipeline, if you don't care about novel isoforms and genes just run Stringtie with -e for each sample, but make sure that the -G option points to the reference transcripts you want estimated (not some merged gtf file generated by StringTie at a previous step).

If you ran it like this please provide some clear examples here about the input references you provided and how the output does NOT have the information you wanted. I can see a potential issue -- currently when run with -e StringTie loses the *gene_name* attribute because it renames it to *ref_gene_name* which we should probably not do for -e (we can fix this easily), but the information is still there, and gene_id and transcript_id are preserved.

**Comment by Michael-w-s** [comment]
Upon further inspection it is not generating novel isoforms as you stated. However, I am getting an error when I run prepDE.py, it was mentioned in another post but I did not see a resolution. The error is: 

Traceback (most recent call last):
  File "/data/kelley/projects/bear_RNA/prepDE.py", line 72, in <module>
    g_id=RE_GENE_ID.search(v[len(v)-1]).group(1)
AttributeError: 'NoneType' object has no attribute 'group'

Do you know what may be causing this?

**Comment by Michael-w-s** [comment]
I have attached two of my sample files, produced by stringtie using the -e and -G flags, and the reference I am using. As an example, if you look at STRG.10188 in OFHi_S30 and STRG.10276 in CFA_S9 they appear to be in the same genomic location and the transcript ID is the same but they have been given different gene IDs. I am not sure why this would be the case. There are many other cases of STRG gene names throughout the samples. I appreciate how quickly you have been responding to questions and I look forward to hearing from you!

[CFA_S9_with_umaritimus_ref.gtf.gz](https://github.com/gpertea/stringtie/files/604249/CFA_S9_with_umaritimus_ref.gtf.gz)
[GCF_000687225.1_UrsMar_1.0_genomic.gff.gz](https://github.com/gpertea/stringtie/files/604250/GCF_000687225.1_UrsMar_1.0_genomic.gff.gz)
[OFHi_S30_with_umaritimus_ref.gtf.gz](https://github.com/gpertea/stringtie/files/604251/OFHi_S30_with_umaritimus_ref.gtf.gz)

**Comment by Michael-w-s** [comment]
I have still not been able to resolve this issue. The gene names are different in separate samples despite appearing to be from the same gene. It seems like this will result in problems with analyzing DE. Have you had a chance to look into this yet? Or am I incorrect and this is how it is intended to work?

**Comment by gpertea** [comment]
I took a quick look and it looks like in those cases StringTie wasn't able to parse a gene_id correctly from the reference annotation data, so it assigned an internal gene ID, which of course is not consistent across samples/runs. StringTie with -e option would never change the transcript_id and the original gene_id -- the exception for the latter being.. when it couldn't find it in the reference annotation. 
So in your example here it's again a problem of the GFF3 parser in StringTie which failed to recognize this feature entry (which corresponds to the transcript ID  _id180629_):
```
NW_007907159.1	Gnomon	C_gene_segment	5489840	5524138	.	-	.	ID=id180629;Parent=gene13907;Dbxref=GeneID:103669001;gbkey=C_region;gene=LOC103669001;model_evidence=Supporting evidence includes similarity to: 2 Proteins%2C and 93%25 coverage of the annotated genomic feature by RNAseq alignments 
```
It looks like the parser couldn't recognize the feature `C_gene_segment` as a transcript (?) so it did not lookup its gene parent in order to retrieve a gene ID -- thus it ended up coming up with its own instead (an internally generate one).

So you are correct, it's not supposed to overwrite the gene IDs, it's a failure of the GFF parser that it couldn't find the parent gene ID for this `C_gene_segment` object.
I'll have to fix this soon..

**Comment by gpertea** [comment]
This should've been fixed in the v1.3.2 release, now the GFF parser should be able to identify any exon-parenting feature as a transcript and fetch the gene_id and gene_name from its parent.

The weird thing about those genes in that file is that their CDS features are actually parented directly by the *gene* (the parent feature of the *X_gene_segment* features), which is quite odd.. Those "gene" features with their CDS will thus be interpreted also as distinct standalone transcripts which have these CDS features overlapping the *exon* features of the *X_gene_segment* features even though they're most likely the same (variable) transcript. Rather messy, but it should not affect much Stringtie's use of such structures as guides, I hope..
